Implement legacy hit explosions for osu!catch

ambeeeeee opened the pull request:
Closes #12095

- [x] Depends on #8019.

Implements hit explosions for legacy skins that closely matches stable's hit explosions.

Comparison video:

https://user-images.githubusercontent.com/4877041/113530354-70c6c680-958b-11eb-8408-8853bb2684a1.mp4

(The hyperfuit objects being incorrect is not relevant to this PR)
The stable catch players I've shown this to are happy with its similarity with stable, and it accomplishes what was asked in the issue. 


Improves pooling logic for catch hit explosions.


Hit explosions for default and legacy currently are not created when the hit object is a Droplet.

ambeeeeee committed:
Make CTB hitlighting skinable and bring in line with other gamemodes.

ambeeeeee committed:
Add documentation, remove test logging statements, and remove chances for null exceptions.

ambeeeeee committed:
Refactor hit explosions interface.
Use PoolableHitExplosion for pooling.
Initial implementation of stable hitexplosion.
Expose more properties to ICatchHitExplosion.

ambeeeeee committed:
Various refactoring and math changes.

ambeeeeee committed:
Attempting to fix transform issues

ambeeeeee committed:
Reduce unexpected transform behaviour.
Adjust the default hitexplosion to better match the current one.
Remove incorrect transformations.

ambeeeeee committed:
Update hit explosion type in catcher hit explosion test.

ambeeeeee committed:
Rename LightingGlow to HitExplosion.

ambeeeeee committed:
Change the initial hitExplosionPool size to its previous value.

ambeeeeee committed:
Renamed HitExplosion.cs to match its classname, DefaultHitExplosion.

Use the correct type when validating hitlighting config.

frenzibyte commented on osu.Game.Rulesets.Catch/UI/Catcher.cs:
```diff
@@ -505,13 +505,15 @@ private Vector2 computePositionInStack(Vector2 position, float displayRadius)
             hitExplosion.ObjectColour = colour;
+            hitExplosion.JudgementResult = result;
+            hitExplosion.CatcherWidth = catchWidth;
+            hitExplosion.CatcherMargin = (1 - ALLOWED_CATCH_RANGE) / 2;
```
sounds like it should just be a constant in `Catcher` or `LegacyCatchHitExplosion`?

frenzibyte commented on osu.Game.Rulesets.Catch/Skinning/Default/FadePiece.cs:
```diff
@@ -0,0 +1,20 @@
+
+namespace osu.Game.Rulesets.Catch.Skinning.Default
+{
+    public class FadePiece : CircularContainer
```
is there a reason this (and `GlowPiece`) got extracted from the `DefaultCatchHitExplosion`?

amberkowalski replied:
I was using osu!standard's hit lighting as a reference and it had all the parts of the hit explosion had their own types. See mainly `osu.Game.Rulesets.Osu/Skinning/Default/FlashPiece.cs`. 

This is just why I did that, and I can just put it back to how it was considering that as of right now both types are just `CircularContainer`s with defaults, rather that implementing any functionality.

bdach replied:
I'm not sure splitting was necessary. These classes are so small I'd just put them back where they were (not to mention that they aren't really properly written - if you wanted to split this to a component, the base class should be `CompositeDrawable` and the `CircularContainer` should be the `InternalChild` here).

frenzibyte reviewed (changes requested):
just a quick review from viewing the changed files, haven't checked thoroughly yet.

amberkowalski commented on osu.Game.Rulesets.Catch/Skinning/Legacy/CatchLegacySkinTransformer.cs:
```diff
@@ -72,6 +72,12 @@ public override Drawable GetDrawableComponent(ISkinComponent component)
                     break;
+
+                case CatchSkinComponents.HitExplosion:
+                    if (GetTexture("scoreboard-explosion-2") != null && GetTexture("scoreboard-explosion-1") != null)
```
It depends, is this intended to validate that we have the textures we need or does it simply check that the skin is a legacy skin? Right now I'm making sure both textures are available for correctness sake.

frenzibyte replied:
The latter, it's intended to check that the skin is a legacy skin, and pass the legacy piece component to display if true. it'll all probably vanish away once falling back to default legacy skin becomes a thing.

amberkowalski replied:
Got it, didn't know that.

ambeeeeee committed:
Bring catch hit explosion construction in line with other skinnables, fix code quality issues, move constant value into its own constant where it is relevant.

Remove CatcherMargin from ICatchHitExplosion and make it a constant in LegacyCatchHitExplosion
Require property getter and setter for all properties in ICatchHitExplosion
Only check if the skin is a legacy skin in CatchLegacySkinTransformer, rather than checking that the textures are available.
LegacyCatchHitExplosion constructor takes ISkinSource instead of textures directly.

Kimitakari commented:
That looks so much better now.

One thing that confuses me is does the hit explosion length matches in stable version because of that video?

bdach commented:
The calculations are the same as stable. The difference can be explained by the difference of the Y position of the catcher.

bdach commented on osu.Game.Rulesets.Catch/Skinning/Legacy/LegacyCatchHitExplosion.cs:
```diff
@@ -0,0 +1,81 @@
+        private readonly Sprite explosion1;
+        private readonly Sprite explosion2;
+
+        /// <inheritdoc cref="Catcher.ALLOWED_CATCH_RANGE"/>
```
We don't really use `inheritdoc` anywhere else, and this one is a touch misleading (mentions "The width of the catcher which can receive fruit", which is inaccurate in this context given the value).

bdach commented on osu.Game.Rulesets.Catch/Skinning/Legacy/LegacyCatchHitExplosion.cs:
```diff
@@ -0,0 +1,81 @@
+        [Resolved]
+        protected ISkinSource Skin { get; private set; }
+
+        public LegacyCatchHitExplosion(ISkinSource source)
+        {
+            Anchor = Anchor.CentreLeft;
+            Origin = Anchor.BottomCentre;
+            RelativeSizeAxes = Axes.Both;
+
+            InternalChildren = new Drawable[]
+            {
+                explosion1 = new Sprite
+                {
+                    Origin = Anchor.CentreLeft,
+                    Texture = source.GetTexture("scoreboard-explosion-2"),
+                    Blending = BlendingParameters.Additive,
+                    Rotation = -90,
+                },
+                explosion2 = new Sprite
+                {
+                    Origin = Anchor.CentreLeft,
+                    Texture = source.GetTexture("scoreboard-explosion-2"),
+                    Blending = BlendingParameters.Additive,
+                    Rotation = -90,
+                }
+            };
```
It is not actually required to pass `Source` in the ctor. It can be done in BDL.

An example of this can be seen [here](https://github.com/ppy/osu/blob/5e427b43844abbfbe40aa4468eebc106e88a457c/osu.Game.Rulesets.Taiko/Skinning/Legacy/TaikoLegacyPlayfieldBackgroundRight.cs#L21-L42).

bdach commented on osu.Game.Rulesets.Catch/Skinning/Legacy/LegacyCatchHitExplosion.cs:
```diff
@@ -0,0 +1,81 @@
+                var scale = Math.Clamp(JudgementResult.ComboAtJudgement / 200f, 0.35f, 1.125f);
+
+                explosion1.Scale = new Vector2(1, 0.9f);
+                explosion1.Alpha = 1;
```
Is this alpha set actually needed? I'd like to think the transform rewind should take care of that.

bdach commented on osu.Game.Rulesets.Catch/Skinning/Legacy/LegacyCatchHitExplosion.cs:
```diff
@@ -0,0 +1,81 @@
+
+            explosion2.ScaleTo(new Vector2(0.9f, 1.3f), 500, Easing.Out).Then().FadeOut(200);
+
+            this.FadeInFromZero().Then().Delay(700).Then().FadeOut(0, Easing.Out);
```
That fade-out at the end doesn't seem to make much sense. It's duration 0, so easing doesn't matter...?

bdach commented on osu.Game.Rulesets.Catch/UI/CatchHitExplosion.cs:
```diff
@@ -0,0 +1,75 @@
+    /// <summary>
+    /// Provides properties hit explosions can use.
+    /// </summary>
+    public interface ICatchHitExplosion
```
I'm not sure how to make sense of this interface. It's a bit weird and I don't think it really needs to exist, given the abstract class should suffice on its own.

amberkowalski replied:
Yeah I was going to bring that up in the change request about adding { get; set; }, I didn't really think I needed the interface at all ever since I added the abstract class. Although the abstract class was added a while after I originally made the interface. I agree.

bdach commented on osu.Game.Rulesets.Catch/UI/Catcher.cs:
```diff
@@ -505,13 +505,14 @@ private Vector2 computePositionInStack(Vector2 position, float displayRadius)
-            HitExplosion hitExplosion = hitExplosionPool.Get();
+            PoolableHitExplosion hitExplosion = hitExplosionPool.Get();
             hitExplosion.HitObject = hitObject;
-            hitExplosion.X = x;
-            hitExplosion.Scale = new Vector2(hitObject.Scale);
+            hitExplosion.CatchPosition = catchPosition;
             hitExplosion.ObjectColour = colour;
+            hitExplosion.JudgementResult = result;
+            hitExplosion.CatcherWidth = catchWidth;
```
This amount of passing data down is disappointing, but in this case I don't see an easy way of resolving this immediately... (especially with `CatchPosition` - which should arguably be in `JudgementResult` - and `CatcherWidth`).

amberkowalski replied:
Additionally not having to pass `CatcherWidth` or `CatchPosition` would simplify DI since I wouldn't need to pass two floats, and thus require one of them to be manually named.

bdach commented on osu.Game.Rulesets.Catch/UI/DefaultHitExplosion.cs:
```diff
@@ -0,0 +1,112 @@
+            }
+            else
+            {
+                this.Delay(50).Then().Delay(duration);
```
What is this supposed to be doing? Is this supposed to delay lifetime end?

amberkowalski replied:
My bad, it was left over from testing.

bdach commented on osu.Game.Rulesets.Catch/UI/DefaultHitExplosion.cs:
```diff
@@ -0,0 +1,112 @@
+
+            const double duration = 400;
+
+            if (!(HitObject is Droplet))
```
This is new (i.e. it wasn't there in the original implementation). Not sure about this change, do players prefer no explosions on droplets? Generally this came out of left field a bit, I don't recall turning explosions off on droplets on the default skin being discussed at any point.

bdach commented on osu.Game.Rulesets.Catch/UI/PoolableHitExplosion.cs:
```diff
@@ -0,0 +1,76 @@
+
+            if (skinnableExplosion?.Drawable is CatchHitExplosion explosion)
+            {
+                explosion.JudgementResult = JudgementResult;
+                explosion.ObjectColour = ObjectColour;
+                explosion.HitObject = HitObject;
+                explosion.CatchPosition = CatchPosition;
+                explosion.CatcherWidth = CatcherWidth;
```
I'd like to get rid of this second layer of data pushdown, if possible. I think the `PoolableHitExplosion` should cache itself so that its children can retrieve it via DI, read the values themselves, and adjust appropriately (probably at the start of `RunAnimation()`?)

amberkowalski replied:
I've added this in 16638a28150c122ac842b3deae1993756c2d4806

bdach commented on osu.Game.Rulesets.Catch/UI/PoolableHitExplosion.cs:
```diff
@@ -0,0 +1,76 @@
+            }
+        }
+
+        public PalpableCatchHitObject HitObject;
+
+        public JudgementResult JudgementResult;
+
+        public float CatcherWidth;
+
+        public float CatchPosition;
```
We generally prefer auto-properties to plain fields. Although it's not a super hard rule.

ambeeeeee committed:
Fix code quality issues and redundant code.

Remove ICatchHitExplosion because CatchHitExplosion has entirely replaced its purpose.
Undo accidental disabling of default hit explosion on Droplets.
Fix parameters on LegacyCatchHitExplosion's FadeOut()
Remove old colour setting code from PoolableHitExplosion.
Add auto properties to PoolableHitExplosion.

ambeeeeee committed:
Remove FadePiece and GlowPiece

ambeeeeee committed:
Use DI to pass data to catch hit explosions.

peppy commented on osu.Game.Rulesets.Catch/Skinning/Legacy/CatchLegacySkinTransformer.cs:
```diff
@@ -72,6 +72,12 @@ public override Drawable GetDrawableComponent(ISkinComponent component)
                     break;
+
+                case CatchSkinComponents.HitExplosion:
+                    if (GetTexture("scoreboard-explosion-2") != null)
```
This check should be based on the presence of a catcher sprite, not the explosion sprite directly. This particular texture is actually not skinnable in `osu-stable` (and we don't want to introduce no skinnable scope where it wasn't there previously) and should always use the default legacy skin implementation.

peppy commented on osu.Game.Rulesets.Catch/Skinning/Legacy/LegacyCatchHitExplosion.cs:
```diff
@@ -0,0 +1,97 @@
+        [Resolved]
+        private Bindable<Color4> objectColour { get; set; }
+
+        [Resolved(Name = "CatcherWidth")]
+        private Bindable<float> catcherWidth { get; set; }
+
+        [Resolved]
+        private Bindable<float> catchPosition { get; set; }
```
We generally don't do this. You'd be better off caching the whole drawable (catcher) or a model class which contains these bindables failing that.

peppy commented on osu.Game.Rulesets.Catch/Skinning/Legacy/LegacyCatchHitExplosion.cs:
```diff
@@ -0,0 +1,97 @@
+        [BackgroundDependencyLoader]
+        private void load(ISkinSource source)
+        {
+            explosion1.Texture = source.GetTexture("scoreboard-explosion-2");
```
As above, these should come from the default legacy skin, rather than the user's skin. I'm not 100% sure we have a method of doing this yet, so may require a bit of further investigation. There's an issue open for the more general case (#8019) but this is an explicit "do not want" case, which is a bit different.

Out of curiosity, does this even work right now? No catch skin should include this sprite as it won't actually apply on osu-stable.

amberkowalski replied:
> Out of curiosity, does this even work right now? No catch skin should include this sprite as it won't actually apply on osu-stable.

You're right about that, I've only been testing with the "classic" skin so it did have the files. I'll take a look at getting it to work as required, then.

peppy replied:
Unfortunately this might not be an easy one, and may need a solution to #8019 as a prerequisite, but see how you go.

peppy commented on osu.Game.Rulesets.Catch/UI/PoolableHitExplosion.cs:
```diff
@@ -0,0 +1,59 @@
+                explosion.RunAnimation();
+            }
+
+            this.Delay(DURATION).Then().Expire(true);
```
Is there a reason this is done with a fixed duration here? I guess this was just copied from mania's implementation?

Wondering if we can potentially use the expiry of the nested skinnable here..

amberkowalski replied:
I can't find a good way to accomplish that, although I agree it should be done. As far as I can tell, there's no method that I can override in `CatchHitExplosion` to accomplish this locally. Provided that I am correct, would adding `OnExpire` or `OnDied` to `Drawable` be something we could do, then hook into it in the pool?

ambeeeeee committed:
Change how values are cached for PoolableHitExplosion.

Cache the entire catcher and make the width public.

Stop caching the hitobject since its available from judgement result.

enoslayd commented:
Can i ask if its intentional that only the classic skin get the hit explosion and not the custom legacy skins?

peppy commented:
@enoslayd see https://github.com/ppy/osu/pull/12293#discussion_r607487895.

peppy commented:
@amberkowalski I was going to look at bringing this one up-to-date, but the merge conflict here is quite complicated. The underlying container which you were modifying has since been updated, to the point it *may* be better to apply your changes again to `master` rather than attempt to get this PR up-to-date.

Are you able to take a look into this? The dependency is almost ready to go, so I think this has a good chance of getting merged in the coming days/weeks.

amberkowalski commented:
Sure, just so we're clear 

> apply your changes again to master 

my thought seeing this is to check out master, then apply all the changes I made in this PR to master manually then rebase and force push. I just want to check since you might mean something else or I might be going about it wrong. Thanks!

peppy commented:
The optimal path is to merge `master` into this branch, but as you'll see it's a bit of a mess.

If you're going to apply changes to `master` please open a new PR rather than force pushing (and reference/close this one).

bdach commented:
As a heads up I've put some work into updating this PR with latest `master` but I still have a few things to check before it's PRable.
